src/snowflake/snowpark/_internal/select_projection_complexity_utils.py

```diff
@@ -0,0 +1,235 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from typing import List
+
+from snowflake.snowpark._internal.analyzer.expression import (
+    Expression,
+    FunctionExpression,
+)
+from snowflake.snowpark._internal.analyzer.table_function import TableFunctionExpression
+from snowflake.snowpark._internal.analyzer.window_expression import WindowExpression
+
+# Set of functions (defined in snowpark-python/src/snowflake/snowpark/functions.py) that do not
+# block merge projections of nested select.
+# Be careful when adding functions in this list, it must statisfy the following condition:
```

Comment:
How do we maintain this list, if in the future we add such a function meeting this requirements? Do we expect developers to add it here

Reply:
That is a good question, yes. right now, I expect developers to add this when they start adding new functions support. the only problem is i could not add instructions to check the function property here because there is an internal parameter needs to be flipped on. Since it requires turn on an internal parameter, we can't do autosync at client side also. I checked the functions satisfies the properties on prod1, there are 673 now, here we covered about 200, which is not bad. Right now, i only added this 200 is more for safety. 

Do we have any place to doc those snowflake internal involved instructions?
The other way we can handle this is like how is_data_generator is added there, also add other properties as part of the FunctionExpression, however, i would like to enforce the configuration of those properties values if we do that. This will continue require developer to explicitly maintain those properties, but might be better since it enforces user think about this when adding new function support. However, the problematic part is about functions.builtin, which is called by snowpark pandas and also open for user usage.  i think we can say if the property is UNKNOWN, we just do not consider it is valid, which may not be valid. 
If we prefer to go through the other way, i think we can do it in a different pr, because it will require touch of a lot of file, and i will need to think about what is the best way to add this since it contains bunch of property.

Reply:
Got it, I'm just not sure how to ensure the developer is aware of this list when adding a new function. Maybe we can add a merge gate like changelog, to check whether any function is added without modifying this list.

Reply:
created ticket to follow up here https://snowflakecomputing.atlassian.net/browse/SNOW-1678506